common/io_lock.c

```diff
@@ -0,0 +1,74 @@
+#include "io_lock.h"
+#include <assert.h>
+#include <ccan/io/io_plan.h>
+
+struct io_lock {
+	bool free;
+};
+
+/* Struct to hold information while we wait for the lock to be freed */
+struct io_lock_waiter {
+	struct io_plan *(*next)(struct io_conn *conn, void *next_arg);
+	void *arg;
+	struct io_lock *lock;
+	enum io_direction dir;
+};
+
+struct io_lock *io_lock_new(const tal_t *ctx)
+{
+	struct io_lock *lock = tal(ctx, struct io_lock);
+	lock->free = true;
+	return lock;
+}
+
+static struct io_plan *io_lock_try_acquire(struct io_conn *conn,
+					   struct io_lock_waiter *waiter)
+{
+	if (waiter->lock->free) {
+		waiter->lock->free = false;
+		return io_out_always(conn, waiter->next, waiter->arg);
+	} else {
+		return io_out_wait(conn, waiter->lock, io_lock_try_acquire, waiter);
```

Comment:
So, internally io knows whether we're dealing with in or out.  Maybe I should simplify the API to avoid this?  I haven't looked hard though, might be a good reason not to...

Reply:
My mistake, I initially had it only for `IO_OUT`, I fixed it up now to check the direction and then calling the correct version. `io_wait` automatically locks `IO_IN` which I think is not correct if we are in `io_duplex` and want to lock the `IO_OUT` explicitly, or am I missing something?